woocommerce/google-listings-and-ads
# API Pull - Add an API to return Google WPCOM App auth URL

### Changes proposed in this Pull Request:



Part of #2146. 

This PR adds an API which returns an auth URL for granting access to Google WPCOM App.

The format of the auth URL:

```
https://public-api.wordpress.com/oauth2/authorize?
client_id=CLIENT_ID
&redirect_uri=PARTNER_REDIRECT_URL
&response_type=code
&blog=BLOD_ID
&scope=wc-partner-access
&state=BASE64_ENCODED_STRING
```

The content of the `state` is an URL query string where the value of its parameter `redirect_url` is being URL encoded. E.g.
`nonce=nonce-123&redirect_url=https%3A%2F%2Fmerchant-site.example.com%2Fwp-admin%2Fadmin.php%3Fpage%3Dwc-admin%26path%3D%2Fgoogle%2Fsetup-mc`

and the decoded version is:

`nonce=nonce-123&redirect_url=https://merchant-site.example.com/wp-admin/admin.php?page=wc-admin&path=/google/setup-mc`

where the `redirect_url` will be used to return back to the merchant's site by the partner.

Project: pcTzPl-1SM-p2
WPCOM App OAuth: https://developer.wordpress.com/docs/oauth2/

### Detailed test instructions:


1. Request to `GET /wc/gla/rest-api/authorize`.
2. Check the returned auth URL has the same format from the description above.
3. Open this URL in the browser, you should see the permission granting page.
4. Make sure you can actually grant the permission and it will be redirected to woo.com.
5. After redirected to woo.com, copy the value of the `state` URL query parameter, which is a base64 encoded string.
6. Go to https://www.base64decode.org/ and decode base64 string.
7. The decoded value should look like
    ```
    nonce=nonce-123&redirect_url=https%3A%2F%2Fmerchant-site.example.com%2Fwp-admin%2Fadmin.php%3Fpage%3Dwc-admin%26path%3D%2Fgoogle%2Fsetup-mc
    ```

### Additional details:

Note that this PR hardcoded three things in the code at the moment as we will be calling a new Google API to get those data when the API is ready:

**Client ID**

WPCOM app Client ID, currently I'm using the WPCOM app I created, it should be Google's client ID in the future.

**Redirect URI**

An URL that Google will set in their WPCOM app settings. I'm currently using https://woo.com set in my WPCOM app.

**nonce**

A validation nonce that would return by the new Google API as well.


### Changelog entry

> Add - API to return Google WPCOM auth URL
